### PR TITLE
quic: add quic version counters in http3 codec stats.

### PR DESCRIPTION
**Pull request by RenjieTang**
Signed-off-by: Renjie Tang <renjietang@google.com>

Commit Message: add quic version counters in http3 codec stats.
Additional Description:
Risk Level: Low
Testing: Integration tests
Docs Changes: docs/root/configuration/http/http_conn_man/stats.rst

### CODE REVIEW

File: docs/root/configuration/http/http_conn_man/stats.rst
Comment (alyssawilk) [inline docs/root/configuration/http/http_conn_man/stats.rst:197]: Hmmm, we have a general Envoy policy that stats don't get removed, but I think it makes sense that as we remove versions of HTTP/3 that we can remove the stats as they'd be implicitly 0 forevermore.

I think this is OK but I want to run this by @envoyproxy/senior-maintainers for second opinion.  Alternately we can not add stats for the versions which are going away, but it'd be annoying between now and Q4 when we remove support for outdated versions.


Second, AFIK (@DavidSchinazi) we plan to remove support for non-H3 QUIC this fall as IETF-quic becomes the standard.  I think we should make this EoL clear here now (even through QUIC is in alpha and has no guarantees) as part of bubbling up data on which versions are supported.

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (alyssawilk) [inline source/common/quic/quic_filter_manager_connection_impl.cc:175]: nit: -> was closed before it could be used 
stats -> Stats

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (alyssawilk) [inline source/common/quic/quic_filter_manager_connection_impl.cc:189]: clean up?

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (alyssawilk) [inline source/common/quic/quic_filter_manager_connection_impl.cc:199]: can we make sure we have unit tests for all of these for coverage?

**Review by alyssawilk — COMMENTED** [review]
Nice addition!  
/wait

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (DavidSchinazi) [inline source/common/quic/quic_filter_manager_connection_impl.cc:162]: In general I'd prefer to plumb ParsedQuicVersion instead of QuicTransportVersion. We currently have a 1:1 mapping but that wasn't always the case and won't necessarily stay that way.

File: docs/root/configuration/http/http_conn_man/stats.rst
Comment (DavidSchinazi) [reply docs/root/configuration/http/http_conn_man/stats.rst:197]: To clarify the plan from the QUIC side of things:
* The default version today is h3-29
* in some number of weeks, h3 (also called RFCv1 in parts of the code) will become the default version, and we plan on supporting that version for years to come
* h3-29 will stay supported for some number of months, but then we'll drop support
* version 43-51 will all be removed in some number of months

I'd suggest only supporting h3/RFCv1 when QUIC leaves Alpha in Envoy, as that's the only version we'll support long-term.

I think we have two options:
1. Add all the current version here, and remove stats as we deprecate versions
1. Add all the current version here, and have old version stats at 0 forever
1. Only add h3-29 and h3 here, and when we remove h3-29 we'll keep that stat at 0 forever

I'd personally lean towards option 1 if that's acceptable to Envoy, but if that's not an option then I'd suggest option 3.

**Review by DavidSchinazi — COMMENTED** [review]
Nice work! Some thoughts inline.

File: docs/root/configuration/http/http_conn_man/stats.rst
Comment (alyssawilk) [reply docs/root/configuration/http/http_conn_man/stats.rst:197]: Yeah, I'd lean towards 1 as well, I just want a thumbs up from a non-googler as a double check.  I think if we comment that the stat will be going away as support is removed by EoY that makes it clear for users.

File: source/common/quic/quic_filter_manager_connection_impl.cc
Comment (RenjieTang) [reply source/common/quic/quic_filter_manager_connection_impl.cc:198]: Sure. I added a new EnvoyClientSessionAllQuicVersionTest for this since none of the existing tests cover all QUIC versions.

File: docs/root/configuration/http/http_conn_man/stats.rst
Comment (RenjieTang) [reply docs/root/configuration/http/http_conn_man/stats.rst:197]: I prefer 1 as well. I want to be able to monitor the QUIC version deprecation process in Envoy as well.
Happy to add more notes in this doc if Envoy maintainers agree.

Comment (alyssawilk) [reply docs/root/configuration/http/http_conn_man/stats.rst:197]: I hear no objections so let's just go for it :-P

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**
